File: android/src/main/java/com/paulingalls/realtimeaudio/RealtimeAudioVADRecorderView.kt
package com.paulingalls.realtimeaudio

import android.content.Context
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher

class RealtimeAudioVADRecorderView(
    context: Context,
    appContext: AppContext
) : BaseAudioView(context, appContext),
    RealtimeAudioVoiceDelegate {
    private val onVoiceCaptured by EventDispatcher<Map<String, String>>()
    private val onVoiceStarted by EventDispatcher()
    private val onVoiceEnded by EventDispatcher()
    private var audioRecorder: RealtimeAudioVADRecorder? = null
    private var isEchoCancellationEnabled = false
    private var modelPath: String = ""
    private var voiceActive: Boolean = false

    override fun setAudioFormat(sampleRate: Int, channelConfig: Int, audioFormat: Int) {
        audioRecorder?.release()
        audioRecorder = RealtimeAudioVADRecorder(sampleRate, channelConfig, audioFormat).apply {
            delegate = this@RealtimeAudioVADRecorderView
            isEchoCancellationEnabled = this@RealtimeAudioVADRecorderView.isEchoCancellationEnabled
            modelPath = this@RealtimeAudioVADRecorderView.modelPath
        }
        super.setAudioFormat(sampleRate, channelConfig, audioFormat)
    }

    fun setEchoCancellationEnabled(echoCancellationEnabled: Boolean) {
        isEchoCancellationEnabled = echoCancellationEnabled
        audioRecorder?.isEchoCancellationEnabled = echoCancellationEnabled
    }

    fun setModelPath(modelPath: String) {
        this.modelPath = modelPath
    }

    fun startListening() {
        running = true
        audioRecorder?.modelPath = modelPath
        audioRecorder?.startListening()
        postInvalidate()
    }

    fun stopListening() {
        running = false
        audioRecorder?.stopListening()
    }

    override fun audioStringReady(base64Audio: String) {
        onVoiceCaptured(mapOf("audioBuffer" to base64Audio))
    }

    override fun voiceBufferReady(buffer: FloatArray) {
        if (!voiceActive) {
            return
        }
        scheduleChunks(buffer)
    }

    override fun voiceStarted() {
        voiceActive = true
        onVoiceStarted(mapOf())
    }

    override fun voiceStopped() {
        onVoiceEnded(mapOf())
        voiceActive = false
        handler.postDelayed({
            visualization.updateData(FloatArray(0))
            audioChunks.clear()
            chunkRenderTimeInMillis = 0
            postInvalidate()
        }, 300)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        audioRecorder?.release()
    }
}

File: ios/RealtimeAudioVADRecorderView.swift
import ExpoModulesCore
import SwiftUI
import AVFoundation

public class RealtimeAudioVADRecorderView: BaseAudioView {
  private var audioRecorder: RealtimeAudioVADRecorder?
  
  let onVoiceCaptured = EventDispatcher()
  let onVoiceStarted = EventDispatcher()
  let onVoiceEnded = EventDispatcher()

  public required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
  }
  
  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
  
  func setAudioFormat(sampleRate: Double, commonFormat: AVAudioCommonFormat, channels: UInt32) {
    audioRecorder = RealtimeAudioVADRecorder(sampleRate: sampleRate, channelCount: channels, audioFormat: commonFormat)
    audioRecorder?.delegate = self
    audioRecorder?.echoCancellationEnabled = echoCancellationEnabled
  }
  
  func startListening() {
    audioRecorder?.startListening()
  }
  
  func stopListening() {
    audioRecorder?.stopListening()
  }
  
  func setEchoCancellationEnabled(_ enabled: Bool) {
    echoCancellationEnabled = enabled
    audioRecorder?.echoCancellationEnabled = enabled
  }
}

extension RealtimeAudioVADRecorderView: RealtimeAudioVADRecorderDelegate {
  func base64VoiceBufferReady(_ base64Audio: String) {
    let event = ["audioBuffer": base64Audio]
    onVoiceCaptured(event)
  }
  
  func voiceBufferCaptured(_ buffer: AVAudioPCMBuffer) {
    updateVisualizationSamples(from: buffer)
  }
  
  func voiceDidStart() {
    onVoiceStarted()
  }
  
  func voiceDidStop() {
    visualization.clearVisualization()
    onVoiceEnded()
  }
}
